### DISCUSSION TITLE
Best way to restart container?

### DISCUSSION THREAD

**Discussion by laneyweb** [lloesche/valheim-server-docker#298] [Q&A]
Hi. I had to restart  (stop and start) the container today. I didnt realise the Valheim server only saved to the DB every 20 mins so lost the last 20 mins of the game :(

What is the best way for me to restart the container but to make sure the game has saved the most recent changes to the DB? 

I see that Supervisor has restart but will that restart the whole container and save the world first? 

Thanks!

**Comment by lloesche** [comment]
If you just `docker restart` it should save the world properly. Make sure to set the stop timeout as described in the README though otherwise docker might run into a timeout (default is 10s) before it had a chance to gracefully shut down the Valheim server and save the world.

**Reply by MapleBZH** [reply]
Hi,
i have update image (no update since a long time.. 2 weeks) and now the `.db` is not save on shutdown.
Backups works fine (with supervisor and cron) but now i must enter `save` in game (admin command) to be sure the .db is save.
What i'm missing ?
I'm using docker UI on a synology NAS and can't change the --stop-timeout option ?

**Reply by lloesche** [reply]
I just pushed an update the might help. It makes sure that valheim-server is the first process to be stopped on shutdown/restart.
That said if you can not adjust the stop-timeout now other processes in the container will be killed hard. But they are likely not stateful.

**Reply by MapleBZH** [reply]
Hi,
thanks ! but still not working :(  should i see "world saved" in log (like every 20min) ?

2021-04-19 11:45:42,646 WARN received SIGTERM indicating exit request
2021-04-19 11:45:42,647 INFO waiting for syslogd, crond, valheim-backup, valheim-server, valheim-updater to die
2021-04-19 11:45:45,651 INFO waiting for syslogd, crond, valheim-backup, valheim-server, valheim-updater to die
2021-04-19 11:45:48,655 INFO waiting for syslogd, crond, valheim-backup, valheim-server, valheim-updater to die
Apr 19 11:45:51 /supervisord: valheim-updater DEBUG - [30] - Received signal to shut down valheim-updater
Apr 19 11:45:51 /supervisord: valheim-updater INFO - Releasing PID file /var/run/valheim-updater.pid
2021-04-19 11:45:51,345 INFO stopped: valheim-updater (exit status 0)
Apr 19 11:45:51 /supervisord: valheim-server DEBUG - [112] - Received signal to shut down valheim-server
Apr 19 11:45:51 /supervisord: valheim-server INFO - Shutting down Valheim server with PID 115
Apr 19 11:45:51 /supervisord: valheim-server DEBUG - [112] - Waiting for Valheim Server with PID 115 to shut down
Apr 19 11:45:51 /supervisord: valheim-server 04/19/2021 11:45:51: Game - OnApplicationQuit
Apr 19 11:45:51 /supervisord: valheim-server 04/19/2021 11:45:51: ZNet Shutdown

**Reply by lloesche** [reply]
That's not the latest container. Looks like Dockerhub was still building when I wrote the answer. I pushed the change 3h ago and Dockerhub's log shows that they only built the image an hour ago. Update to the latest version and you should see that `valheim-server` is the first thing that gets shut down.

Upon start it should say
```
Apr 19 10:08:18 /supervisord: valheim-bootstrap DEBUG - [27] - Running commit 8239e589d3ccdb2b40b77b41c3cb0c6d401c30e3
```

That's the latest version currently.

**Reply by MapleBZH** [reply]
OK thanks, i have now "clone" in the log!! but my server is too slow i think, when i manually save it takes 8sec... :

2021-04-19 12:17:23,290 WARN received SIGTERM indicating exit request
2021-04-19 12:17:23,291 INFO waiting for syslogd, crond, valheim-backup, valheim-updater, valheim-server to die
Apr 19 12:17:23 /supervisord: valheim-server DEBUG - [108] - Received signal to shut down valheim-server
Apr 19 12:17:23 /supervisord: valheim-server INFO - Shutting down Valheim server with PID 111
Apr 19 12:17:23 /supervisord: valheim-server DEBUG - [108] - Waiting for Valheim Server with PID 111 to shut down
Apr 19 12:17:23 /supervisord: valheim-server 04/19/2021 12:17:23: Game - OnApplicationQuit
Apr 19 12:17:23 /supervisord: valheim-server 04/19/2021 12:17:23: ZNet Shutdown
2021-04-19 12:17:26,366 INFO waiting for syslogd, crond, valheim-backup, valheim-updater, valheim-server to die
**Apr 19 12:17:28 /supervisord: valheim-server 04/19/2021 12:17:28: clone 4772**
Apr 19 12:17:29 /supervisord: valheim-server DEBUG - [108] - Waiting for Valheim Server with PID 111 to shut down
2021-04-19 12:17:30,309 INFO waiting for syslogd, crond, valheim-backup, valheim-updater, valheim-server to die
_and nothing else after_

Test with a "light" world :

2021-04-19 12:24:40,992 WARN received SIGTERM indicating exit request
2021-04-19 12:24:40,993 INFO waiting for syslogd, crond, valheim-backup, valheim-updater, valheim-server to die
Apr 19 12:24:40 /supervisord: valheim-server DEBUG - [106] - Received signal to shut down valheim-server
Apr 19 12:24:40 /supervisord: valheim-server INFO - Shutting down Valheim server with PID 109
Apr 19 12:24:41 /supervisord: valheim-server DEBUG - [106] - Waiting for Valheim Server with PID 109 to shut down
Apr 19 12:24:41 /supervisord: valheim-server 04/19/2021 12:24:41: Game - OnApplicationQuit
Apr 19 12:24:41 /supervisord: valheim-server 04/19/2021 12:24:41: ZNet Shutdown
**Apr 19 12:24:42 /supervisord: valheim-server 04/19/2021 12:24:42: clone 1514**
2021-04-19 12:24:44,434 INFO waiting for syslogd, crond, valheim-backup, valheim-updater, valheim-server to die
**Apr 19 12:24:44 /supervisord: valheim-server 04/19/2021 12:24:44: Saved 307332 zdos**
**Apr 19 12:24:44 /supervisord: valheim-server 04/19/2021 12:24:44: World saved ( 2056.619ms )**
Apr 19 12:24:45 /supervisord: valheim-server 04/19/2021 12:24:45: Sending disconnect msg
Apr 19 12:24:45 /supervisord: valheim-server 04/19/2021 12:24:45: Disposing socket
Apr 19 12:24:45 /supervisord: valheim-server 04/19/2021 12:24:45: Stopping listening socket
Apr 19 12:24:45 /supervisord: valheim-server 04/19/2021 12:24:45: Last socket, unregistering callback
Apr 19 12:24:45 /supervisord: valheim-server 04/19/2021 12:24:45: ZSteamSocket  UnregisterGlobalCallbacks, existing sockets:0
Apr 19 12:24:47 /supervisord: valheim-server DEBUG - [106] - Waiting for Valheim Server with PID 109 to shut down
Apr 19 12:24:47 /supervisord: valheim-server Setting up 1 worker threads for Enlighten.
Apr 19 12:24:47 /supervisord: valheim-server   Thread -> id: 7fcbfffbb700 -> priority: 1 
2021-04-19 12:24:48,133 INFO waiting for syslogd, crond, valheim-backup, valheim-updater, valheim-server to die
Apr 19 12:24:48 /supervisord: valheim-server 04/19/2021 12:24:48: Steam manager on destroy
Apr 19 12:24:49 /supervisord: valheim-server 04/19/2021 12:24:49: ZNet OnDestroy
Apr 19 12:24:49 /supervisord: valheim-server 04/19/2021 12:24:49: Net scene destroyed

So considered it is solve you can't do anything with slow hardware ;)
Thank you for your amazing work !

**Reply by lloesche** [reply]
You could try if you can ssh into the Synology and increase the stop-timeout manually. Docker has the ability. I'm guessing Synology just didn't include it with the Web UI.

**Reply by MapleBZH** [reply]
works fine in SSH thanks!
